[infra-proxy-service] Manage node tags with add, delete and set actions

vsingh-msys opened the pull request:
### :nut_and_bolt: Description: What code changed, and why?

<!-- /!\ Please ensure that you are NOT disclosing any customer information without their consent /!\ -->

In order to manage the tags, we have three sets of action as follows:

 - `add`: Adds the tags to existing node tags.
 - `delete`: Deletes the tags present on the node.
 - `set`: Reset the tags with given tags for the node.

Also did some code-refactoring for `UpdateNode` and `DeleteNode` proto request format. These APIs under development so no impact on proto changes. [02/22/2020 Updated]

### :chains: Related Resources
Fixes: #4530 

### :+1: Definition of Done

- Infra service is able to manage the node tags using add, delete, and set operations.

### :athletic_shoe: How to Build and Test the Change

Steps to build:
 - `rebuild components/infra-proxy-service`
 - `rebuild components/automate-gateway`

Steps to test:

- To add data using 3rd internal steps https://github.com/chef/automate/blob/master/dev-docs/adding-data/adding_test_data.md#adding-data-to-infra-views

- Node tags add API:

```
 curl -sSkH "api-token: $(get_admin_token)" -X PUT https://a2-dev.test/api/v0/infra/servers/chef-server-dev-test/orgs/chef-org-ev/nodes/aix-7.1/tags?pretty  -d '{"action": "add", "tags": ["tag1"] }'
{
  "tags": [
    "aix7.1",
    "aix",
    "ibm",
    "tag1"
  ]
}
```

- Node tags remove API:
```
curl -sSkH "api-token: $(get_admin_token)" -X PUT https://a2-dev.test/api/v0/infra/servers/chef-server-dev-test/orgs/chef-org-ev/nodes/aix-7.1/tags?pretty  -d '{"action": "delete", "tags": ["tag1"] }'
{
  "tags": [
    "aix7.1",
    "aix",
    "ibm"
  ]
}
```

- Node tags set API:
```
 curl -sSkH "api-token: $(get_admin_token)" -X PUT https://a2-dev.test/api/v0/infra/servers/chef-server-dev-test/orgs/chef-org-ev/nodes/aix-7.1/tags?pretty  -d '{"action": "set", "tags": ["ibm"] }'
{
  "tags": [
    "ibm"
  ]
}
```

### :white_check_mark: Checklist

- [x] I have read the [CONTRIBUTING document](https://github.com/chef/automate/blob/master/CONTRIBUTING.md).
- [x] Tests added/updated?
- [ ] Docs added/updated?
- [x] All commits have been signed-off for the [Developer Certification of Origin](https://github.com/chef/chef/blob/master/CONTRIBUTING.md#developer-certification-of-origin-dco).

### :camera: Screenshots, if applicable
Signed-off-by: Vivek Singh <vivek.singh@msystechnologies.com>

punitmundra commented on components/infra-proxy-service/server/nodes.go:
```diff
@@ -191,6 +191,56 @@ func (s *Server) UpdateNode(ctx context.Context, req *request.UpdateNode) (*resp
+		tags = Unique(append(tags, req.Tags...))
+	case "delete":
+		tags = SubstractSlice(tags, req.Tags)
+	case "set":
```
does add will do append the remaining list and set will create the new set of tag ?

vsingh-msys replied:
Yes, add will be used for appending to existing tags, set will reset/create a new set of tags.

vsingh-msys commented on api/external/infra_proxy/infra_proxy.proto:
```diff
@@ -255,12 +255,12 @@ service InfraProxy {
     option (chef.automate.api.iam.policy).action = "infra:infraServers:update";
   };
-  rpc UpdateNode (infra_proxy.request.UpdateNode) returns (infra_proxy.response.UpdateNode) {
+  rpc UpdateNode (infra_proxy.request.NodeObject) returns (infra_proxy.response.Node) {
```
`UpdateNode` & `CreateNode` now using same request `NodeObject` so we can reuse it.

vsingh-msys commented on api/external/infra_proxy/request/nodes.proto:
```diff
@@ -34,16 +34,7 @@ message AffectedNodes {
     string version = 5;
 }
 
-message DeleteNode {
-    // Chef organization ID.
-    string org_id = 1;
-    // Chef Infra Server ID.
-    string server_id = 2;
-    // Node name.
-    string name = 3;
-}
```
Use `Node` in place of `DeleteNode` api.

vsingh-msys commented on components/infra-proxy-service/server/nodes.go:
```diff
@@ -191,6 +191,56 @@ func (s *Server) UpdateNode(ctx context.Context, req *request.UpdateNode) (*resp
+		tags = Unique(append(tags, req.Tags...))
+	case "delete":
+		tags = SubstractSlice(tags, req.Tags)
+	case "set":
```
Added required integration tests https://github.com/chef/automate/blob/VSingh/infra-proxy-service-manage-node-tags/components/infra-proxy-service/integration_test/nodes_test.go

vsingh-msys commented:
- Added `nodes_test.go` file to cover the integration test cases. Also added interservice `CreateNode` for adding the test cases.


This PR is ready for re-review please have look.

msorens commented on components/infra-proxy-service/integration_test/nodes_test.go:
```diff
@@ -0,0 +1,152 @@
+		assert.Equal(t, []string{"tag1", "tag2", "tag3"}, res.Tags)
+	})
+
+	t.Run("when set tags request with valid params", func(t *testing.T) {
```
```suggestion
	t.Run("when set, replaces tags with specified values", func(t *testing.T) {
```

msorens commented on components/infra-proxy-service/integration_test/nodes_test.go:
```diff
@@ -0,0 +1,152 @@
+
+func TestUpdateNodeTags(t *testing.T) {
+	ctx := context.Background()
+	t.Run("when add tags request with valid params", func(t *testing.T) {
```
```suggestion
	t.Run("when add, adds values to the current group of tags", func(t *testing.T) {
```

msorens commented on components/infra-proxy-service/integration_test/nodes_test.go:
```diff
@@ -0,0 +1,152 @@
+		assert.Equal(t, []string{"reset1"}, res.Tags)
+	})
+
+	t.Run("when delete tags request with valid params", func(t *testing.T) {
```
```suggestion
	t.Run("when delete, removes tags from the current group of values", func(t *testing.T) {
```

msorens commented on components/infra-proxy-service/integration_test/nodes_test.go:
```diff
@@ -0,0 +1,152 @@
+	})
+}
+
+func TestUpdateNodeTags(t *testing.T) {
```
Additional tests suggested:
 * when attempting to delete a non-existent tag, leaves tag group unchanged (?? or throws an error, whatever is correct here)
 * when attempting to delete all tags, updates tag group to the empty list
 * when attempting to delete some existing and some non-existing tags simultaneously, does ??
 * when attempting to set tags to empty list, updates tag group to the empty list

msorens commented on components/infra-proxy-service/server/util.go:
```diff
@@ -39,3 +39,38 @@ func SafeSliceFromMap(values map[string]interface{}, key string) []string {
 	return []string{}
 }
+
+// SubstractSlice return the slice without containing all element of b.
+func SubstractSlice(a []string, b []string) []string {
```
Fix typo and make the explanation more clear.
```suggestion
// SubtractSlice returns the slice a with all elements of b removed.
func SubtractSlice(a []string, b []string) []string {
```

msorens commented on components/infra-proxy-service/server/util_test.go:
```diff
@@ -44,3 +44,28 @@ func TestSafeSliceFromMap(t *testing.T) {
 	assert.Equal(t, []string{}, SafeSliceFromMap(testData, "no-key"))
 }
+
+func TestSubstractSlice(t *testing.T) {
```
Add tests:
* subtract both tag3 and unknown from the list of tag1, tag2, and tag3.
* subtract empty list from a non-empty list
* subtract a non-empty list from an empty list
* subtract whole list from itself

vsingh-msys replied:
Thank you for providing much detailed test cases scenario!

msorens commented on components/infra-proxy-service/server/util_test.go:
```diff
@@ -44,3 +44,28 @@ func TestSafeSliceFromMap(t *testing.T) {
+	assert.Equal(t, testData3, RemoveElement(testData3, ""))
+}
+
+func TestUnique(t *testing.T) {
```
Add tests:
 * tag1, tag2, tag3, tag1 (i.e. non-adjacent in the list)
 * tag, tag, tag (i.e. all elements are the same)
 * empty list

msorens commented on components/infra-proxy-service/server/util_test.go:
```diff
@@ -44,3 +44,28 @@ func TestSafeSliceFromMap(t *testing.T) {
+	assert.Equal(t, []string{"tag1", "tag2", "tag3"}, SubstractSlice([]string{"tag1", "tag2", "tag3"}, []string{"unknown"}))
+}
+
+func TestRemoveElement(t *testing.T) {
```
Add tests:
* Remove value from empty list
* Remove empty string from empty list

vsingh-msys commented on components/infra-proxy-service/integration_test/nodes_test.go:
```diff
@@ -0,0 +1,152 @@
+		}
+
+		res, err := infraProxy.GetNode(ctx, req)
+		assert.Error(t, err)
```
oh yeah, forgot to add this, thanks for pointing it out.

vsingh-msys commented:
@msorens Thank you for reviewing PR in detail, Incorporated the reviews comments and updated the PR.  

Some of the reviews could be minimized from my side, will take this into consideration for further PRs Thanks!

msorens reviewed (approved):
Nice work, @vsingh-msys !
